crates/revmc-worker/src/external.rs

```diff
@@ -16,53 +23,72 @@ pub(crate) static SLED_DB: OnceCell<Arc<RwLock<SledDB<B256>>>> = OnceCell::new()
+            cache: RwLock::new(LruCache::new(NonZeroUsize::new(cache_size_words).unwrap())),
+        }
     }
 
-    pub fn get_function(&mut self, code_hash: B256) -> Result<Option<EvmCompilerFn>, ExtError> {
+    /// Fetches the compiled function from disk, if exists
+    ///
+    /// Does not utilize EXT for Address Zero
+    /// When parallel compilation is in progress for the same code_hash,
+    /// it resumes without utilizing the cached ExternalFn
+    pub fn get_function(&self, code_hash: B256) -> Result<Option<EvmCompilerFn>, ExtError> {
         if code_hash.is_zero() {
             return Ok(None);
         }
 
-        if let Some((f, _)) = self.cache.get(&code_hash) {
-            return Ok(Some(*f));
+        // Counter-intuitively, Write locks are required for reading from LRU Cache
+        {
+            let mut write_lock = match self.cache.try_write() {
+                Ok(g) => g,
+                Err(err) => match err {
+                    // Compile in progress, resume without EXT
+                    TryLockError::WouldBlock => {
+                        return Ok(None);
+                    }
+                    TryLockError::Poisoned(err) => {
+                        return Err(ExtError::RwLockPoison { err: err.to_string() });
+                    }
+                },
+            };
+
+            if let Some((f, _)) = write_lock.get(&code_hash) {
+                return Ok(Some(*f));
+            }
         }
 
         let so_file = aot_store_path().join(code_hash.to_string()).join("a.so");
         let exist: bool = so_file.try_exists().unwrap_or(false);
         if exist {
             {
-                let lib = (unsafe { libloading::Library::new(&so_file) }).map_err(
-                    |err| ExtError::LibLoadingError { err: err.to_string() }
-                )?;
+                let lib = (unsafe { libloading::Library::new(&so_file) })
+                    .map_err(|err| ExtError::LibLoadingError { err: err.to_string() })?;
+
                 let f: EvmCompilerFn = unsafe {
-                    *lib
-                        .get(code_hash.to_string().as_ref())
+                    *lib.get(code_hash.to_string().as_ref())
                         .map_err(|err| ExtError::GetSymbolError { err: err.to_string() })?
                 };
+
                 // The function holds a reference to the library, so dropping the library will cause an error.
                 // Therefore, the library must also be stored in the cache.
-                self.cache.put(code_hash, (f, lib));
+                let mut write_lock = match self.cache.write() {
+                    Ok(g) => g,
+                    Err(err) => return Err(ExtError::RwLockPoison { err: err.to_string() }),
+                };
+                write_lock.put(code_hash, (f, lib));
 
-                if let Some((f, _)) = self.cache.get(&code_hash) {
+                if let Some((f, _)) = write_lock.get(&code_hash) {
```

Comment:
we don't have to get function from cache. just return f

```suggestion
    return Ok(Some(*f));
```

Reply:
the purpose of this is to check if function exists in cache

Reply:
if there are any cases of putting f, lib in cache? if there are not, i think we don't have to check it